### DISCUSSION TITLE
tutorials/lift-off-part3/04-resolver-args-parameter

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#19] [course feedback]
# tutorials/lift-off-part3/04-resolver-args-parameter

https://www.apollographql.com/tutorials/lift-off-part3/04-resolver-args-parameter

**Comment by manavm1990** [comment]
```
// Here's the solution (your code is commented out after):
const resolvers = {
  Query: {
    spaceCat: (_, { id }, { dataSources } ) => {
      return dataSources.spaceCatsAPI.getSpaceCat(id);
    }
  }
}
/* your code:
const resolvers = {
  Query: {
    // write your resolver here
    spacecat: (_, {id}, {dataSources}) => {
      return dataSources.spaceCatsAPI.getSpaceCat(id)
    }
  }
}
*/
```

It's same, no?

**Reply by manavm1990** [reply]
Ugh! NVM. I see it. 🐛

**Reply by mabuyo** [reply]
:cat:

**Reply by Rujal-Duwal** [reply]
I cant see the difference?
what is the issue here?

**Reply by R4ph-t** [reply]
Just a little typo (which would result in an error though). 
The casing on the spaceCat Query was not the expected one on the submitted answer. It's supposed to be camelCase: `spaceCat` capital C and not `spacecat`

**Comment by astroecom** [comment]
Hi, mabuyo. nice to talk to you. I 've a problem with argument of resolve function.
May I update track resolver like this to resolve.js?
    track:({trackId}, arg, { dataSources } ) => {
      return dataSources.trackAPI.getTrack(trackId);
    }

As you mentioned in list-off-2 tutorial, authorId of author resolve function could get from parent of resolve function argument.
Why should I get this trackId in track resolve function from arg?

**Reply by neogeek13** [reply]
I was wondering the same thing...

**Comment by astroecom** [comment]
Oh, sorry... I saw your video clips again. I misunderstood track and author resolve function level.  Track is parent of author data layer, right?
And then may I write code like this?
track:(_, {trackId}, { dataSources } ) => {
return dataSources.trackAPI.getTrack(trackId);
}

There is only one difference with ur code in id and trackId. Please tell me about it.

**Reply by lizhennessy** [reply]
Hi there @BenDev1001!

The names of the arguments we can expect to receive on a field's `args` parameter depend on how we have implemented that particular field in the schema. If you recall, in Lesson 3: GraphQL arguments, we set up a `track` field on our schema's `Query` type, and gave it an argument with the name of `id`.

```
type Query {
     # ... other fields
     "Fetch a specific track, provided a track's ID"
     track(id: ID!): Track
}
```

For this reason, when we query for the `track` field, we need to pass it an argument called `id`. This argument is included on the `args` parameter that reaches the `track` field's resolver function.

```
// get a single track by ID, for the track page
track: (_, {id}, {dataSources}) => {
  return dataSources.trackAPI.getTrack(id);
},
```

In order for `trackId` to work as the name `track` field's argument, we would have needed to specify `trackId` as one of the arguments the `track` field accepts, as the example below shows - notice that I've replaced `id` with `trackId`. 

```
type Query {
     # EXAMPLE
    track(trackId: ID!): Track
}
```

You'll notice that the `Track` object type also has a field called `id` and not `trackId`; we chose to use `id` because as a field belonging to the `Track` type, we can assume that the `id` value is relevant to a specific track, and not a module, for instance. This also lets us match the format that our track data might be stored in a database. You'll see in the resolver function that with an argument called `id`, we're able to pass the value directly into our `trackAPI` data source's `getTrack` method. It understands just how to retrieve the track we're looking for when we provide it with a specific value.

I hope that this helps answer your question! Please reach out if there's anything else we can do to assist.

**Reply by astroecom** [reply]
Thanks very much.

**Comment by PoojaMgr** [comment]
Hi there! My confusion is still not clear why authorId of author resolve function could get from parent of resolve function argument but this trackId in track resolve function from arg?

**Reply by mabuyo** [reply]
Hi there @PoojaMgr ! 

For the `Query.track(id: ID!)` resolver, `id` is available in the field's arguments.

```graphql
type Query {
  track(id: ID!): Track  # id is available in arguments
}
```

Comparing with `Track.author`, there are no arguments in that field to access the track ID. So we can't use `args`. Instead, we need to access the parent object (`Track`) to retrieve the ID.

```graphql
type Track {
  id: ID!  # We want this value, which is accessible through the `parent`
  author: Author
}
```

This is what the resolver chain looks like:

`Query.track()` -> returns `Track`, which becomes the `parent` to -> `Track.author()`

In this case, the returned `Track` object contains the track's `id` field, which the `Track.author` resolver can access through its `parent` object. Hope that helps!
